Sources/CredentialsFacebook/TypeSafeFacebook.swift

```diff
@@ -0,0 +1,159 @@
+/**
+ * Copyright IBM Corporation 2018
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ **/
+
+import Foundation
+import KituraNet
+import Credentials
+import LoggerAPI
+import TypeDecoder
+
+/// A protocol that defines common attributes of Facebook authentication methods (token
+/// and redirecting authentication).
+///
+/// It is not intended for a user's type to conform to this protocol directly. Instead,
+/// your type should conform to a specific authentication type, such as
+/// `TypeSafeFacebookToken`.
+public protocol TypeSafeFacebook: TypeSafeCredentials {
+    /// The OAuth client id ('AppID') that tokens should correspond to. This value must be
+    /// set to match the Facebook OAuth app that was used to issue the token. Tokens that
+    /// are received but that do not match this value will be rejected.
+    static var appID: String { get }
+
+    /// A set of valid field names that can be requested from Facebook. A default set is
+    /// implemented for you, however this property can be overridden if needed to customize
+    /// or extend the set.
+    static var validFieldNames: Set<String> { get }
+
+}
+
+extension TypeSafeFacebook {
+    /// Provides a default provider name of `Facebook`.
+    public var provider: String {
+        return "Facebook"
+    }
+
+    /// Defines the list of valid fields that can be requested from Facebook.
+    /// Source: https://developers.facebook.com/docs/facebook-login/permissions/v3.0#reference-extended-profile
```

Comment:
Why do we link to #reference-extended-profile? It is not easy to see from the facebook page you reference here which are the field names that we have provided default implementations for as the API doc doesn't show that.  Are we expecting users to refer to a specific type such as `TypeSafeFacebookToken` to find out which are the valid field types?

Reply:
Whoops, good catch! This was meant to be https://developers.facebook.com/docs/facebook-login/permissions/v3.0#reference-default_fields